src/device_trezor/trezor/protocol.cpp

```diff
@@ -0,0 +1,1074 @@
+//    materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+#include "protocol.hpp"
+#include <unordered_map>
+#include <set>
+#include <utility>
+#include <boost/endian/conversion.hpp>
+#include <common/apply_permutation.h>
+#include <ringct/rctSigs.h>
+#include <ringct/bulletproofs.h>
+#include "sodium.h"
+#include "sodium/crypto_aead_chacha20poly1305.h"
+
+namespace hw{
+namespace trezor{
+namespace protocol{
+
+#define BULLETPROOF_MAX_OUTPUTS 16
+
+  std::string key_to_string(const ::crypto::ec_point & key){
+    return std::string(key.data, 32);
+  }
+
+  std::string key_to_string(const ::crypto::ec_scalar & key){
+    return std::string(key.data, 32);
+  }
+
+  std::string key_to_string(const ::crypto::hash & key){
+    return std::string(key.data, 32);
+  }
+
+  std::string key_to_string(const ::rct::key & key){
+    return std::string(reinterpret_cast<const char*>(key.bytes), 32);
+  }
+
+  void string_to_key(::crypto::ec_scalar & key, const std::string & str){
+    if (str.size() != 32){
+      throw std::invalid_argument("Key has to have 32 B");
+    }
+    memcpy(key.data, str.data(), 32);
+  }
+
+  void string_to_key(::crypto::ec_point & key, const std::string & str){
+    if (str.size() != 32){
+      throw std::invalid_argument("Key has to have 32 B");
+    }
+    memcpy(key.data, str.data(), 32);
+  }
+
+  void string_to_key(::rct::key & key, const std::string & str){
+    if (str.size() != 32){
+      throw std::invalid_argument("Key has to have 32 B");
+    }
+    memcpy(key.bytes, str.data(), 32);
+  }
+
+namespace crypto {
+
+    int ct_equal(const char *a, const char *b, size_t len) {
+    size_t i;
+    unsigned int dif = 0;
+    for (i = 0; i < len; i++)
+      dif |= (a[i] ^ b[i]);
+    dif = (dif - 1) >> ((sizeof(unsigned int) * 8) - 1);
+    return (dif & 1);
+  }
+
+namespace chacha {
+
+  void decrypt(const void* data, size_t length, const uint8_t* key, const uint8_t* iv, char* cipher){
```

Comment:
Bikeshedding, but it's a decrypt function, the output isn't a ciphertext, but a plaintext :)